tachyon/crypto/sumcheck/sumcheck_prover.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_proving_key.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/math/polynomials/multivariate/linear_combination.h"
+
+namespace tachyon::crypto {
+
+// The prover argues for the sum of the polynomial over {0,1}^|num_variables_|
+//
+// The polynomial is represented by a list of terms each containing a
+// coefficient and a list of |MultilinearDenseEvaluations|. Refer to
+// "tachyon/math/polynomials/multivariate/linear_combination.h" for more info.
+template <typename MLE>
+class SumcheckProver {
+ public:
+  using F = typename MLE::Field;
+  using Term = math::LinearCombinationTerm<F>;
+  using Point = std::vector<F>;
+
+  SumcheckProver() = default;
+  explicit SumcheckProver(const SumcheckProvingKey<MLE>& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(key.terms),
+        flattened_ml_evaluations_(key.flattened_ml_evaluations) {
+    // Constants are not accepted
+    CHECK_NE(num_variables_, size_t{0});
+    randomness_.reserve(num_variables_);
+  }
+
+  // Generate prover message, and proceed to next round
+  //
+  // Main algorithm used is from section 3.2 of
+  // [XZZPS19](https://eprint.iacr.org/2019/317.pdf#subsection.3.2).
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // For v = the number of variables...
+  // For the first round, the Prover creates and sends:
+  //   g₁(X₁) := ∑ g(X₁,x₂,...,xᵥ), (x₂,...,xᵥ)∈{0,1}ᵛ⁻¹
+  // For the consequent round i, the Prover creates and sends:
+  //   gᵢ(Xᵢ) = ∑ g(r₁,...,rᵢ₋₁,Xᵢ,xᵢ₊₁,...,xᵥ), (xᵢ₊₁ ,...,xv)∈{0,1}ᵛ⁻¹
+  // More specifically, |Prover(VerifierMsg)| fixes the front variables with
+  // r₁,...,rᵢ₋₁, and |Prover()| creates the total gᵢ(Xᵢ)
```

Comment:
If it's for the first round, why don't you remove those or move somewhere else?

Reply:
`Round()` runs for all Prover rounds including the first round, so it's not limited to the first round. If a Prover round creates this every time: gᵢ(Xᵢ) = ∑ g(r₁,...,rᵢ₋₁,Xᵢ,xᵢ₊₁,...,xᵥ), `Round()` creates ∑ g(Xᵢ,xᵢ₊₁,...,xᵥ) and `Round(VerifierMsg)` fixes variables (∑ g(r₁,...,rᵢ₋₁,Xᵢ,xᵢ₊₁,...,xᵥ) <- "r" parts) and then runs `Round()` to compute the whole poly.